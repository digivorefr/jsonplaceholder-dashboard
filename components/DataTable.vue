<template>
  <div class="data-table">
    <div class="data-table__scroller">
      <div class="data-table__wrapper">
        <div
          v-for="columnIndex in columns"
          :key="columnIndex"
          class="data-table__column"
        >
          <div
            v-if="headers !== null"
            class="data-table__cell data-table__cell--heading"
          >
            {{ headers[columnIndex -1] }}
          </div>
          <DataTableCell
            v-for="(row, index) in values"
            :key="index"
            :value="row[columnIndex-1]"
          />
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import Vue, { VNode } from 'vue'
import DataTableCell, { Value } from '~/components/DataTableCell.vue'
// export type Value = string | VNode | null;
export type Action = VNode | null;

export type Options = {
  values: Value[][],
  headers?: string[],
}

export default Vue.extend({
  name: 'DataTable',
  components: { DataTableCell },
  props: {
    options: {
      type: Object as () => Options,
      required: true
    }
  },
  computed: {
    headers(): string[] | null {
      if (this.options.headers !== undefined &&
        this.options.headers !== null &&
        this.options.headers.length !== this.options.values[0].length) {
        this.$nuxt.error({ statusCode: 500, message: 'DataTable headers and values must have the same length.' })
      }
      return this.options.headers || null
    },
    values(): Value[][] {
      return this.options.values
    },
    columns(): number { return (this.values[0] || []).length },
    rows(): number { return this.values.length }
  }
})
</script>
