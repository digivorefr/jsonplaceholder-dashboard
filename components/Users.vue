<template>
  <div>
    <h1>Users</h1>
    <Loader v-if="status === 'loading'" />
    <DataTable
      v-else-if="users.length > 0"
      :options="buildTableData()"
    />
    <div v-else>
      No users.
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import Loader from '~/components/Loader.vue'
import { User, Status } from '~/store'
import DataTable, { Options as DataTableOptions } from '~/components/DataTable.vue'
import { Value } from '~/components/DataTableCell.vue'

export default Vue.extend({
  name: 'UsersView',
  components: { Loader, DataTable },
  computed: {
    users(): User[] {
      return this.$store.getters.getUsers
    },
    status(): Status {
      return this.$store.state.status as Status
    }
  },
  methods: {
    buildTableData(): DataTableOptions {
      return {
        headers: ['#', 'Username', 'Email', 'Company'],
        values: this.users.map(user => [
          `${user.id}`,
          {
            component: 'nuxt-link',
            props: {
              to: `/users/${user.id}`,
              children: `${user.username}`
            }
          },
          user.email,
          user.company.name
        ] as Value[])
      }
    }
  }
})
</script>
