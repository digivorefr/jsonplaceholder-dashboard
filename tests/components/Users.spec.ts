import { shallowMount, Wrapper } from '@vue/test-utils'
import Users from '@/components/Users.vue'
import DataTable from '@/components/DataTable.vue'

const storeMock = {
  getters: {
    getUsers: [
      {
        id: 1,
        name: 'Leanne Graham',
        username: 'Bret',
        email: 'Sincere@april.biz',
        address: {
          street: 'Kulas Light',
          suite: 'Apt. 556',
          city: 'Gwenborough',
          zipcode: '92998-3874',
          geo: {
            lat: '-37.3159',
            lng: '81.1496'
          }
        },
        phone: '1-770-736-8031 x56442',
        website: 'hildegard.org',
        company: {
          name: 'Romaguera-Crona',
          catchPhrase: 'Multi-layered client-server neural-net',
          bs: 'harness real-time e-markets'
        }
      },
      {
        id: 2,
        name: 'Ervin Howell',
        username: 'Antonette',
        email: 'Shanna@melissa.tv',
        address: {
          street: 'Victor Plains',
          suite: 'Suite 879',
          city: 'Wisokyburgh',
          zipcode: '90566-7771',
          geo: {
            lat: '-43.9509',
            lng: '-34.4618'
          }
        },
        phone: '010-692-6593 x09125',
        website: 'anastasia.net',
        company: {
          name: 'Deckow-Crist',
          catchPhrase: 'Proactive didactic contingency',
          bs: 'synergize scalable supply-chains'
        }
      }
    ]
  },
  state: {
    status: 'loaded'
  }
}

describe('Users', () => {
  test('renders correctly', async () => {
    const wrapper: Wrapper<Vue & { [key: string]: any }> = shallowMount(Users, {
      mocks: {
        $store: storeMock
      }
    })
    await wrapper.vm.$nextTick()
    expect(wrapper.vm).toBeTruthy()
    expect(wrapper.exists()).toBe(true)
    expect(wrapper.findComponent(DataTable)).toBeTruthy()
    expect(wrapper.vm.users.length).toBe(2)
    expect(wrapper.element).toMatchSnapshot()
    wrapper.destroy()
  })
})
