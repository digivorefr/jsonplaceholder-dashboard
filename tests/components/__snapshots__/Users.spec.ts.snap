// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users renders correctly 1`] = `
<div>
  <h1>
    Users
  </h1>
   
  <datatable-stub
    options="[object Object]"
  />
</div>
`;
